flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java

```diff
@@ -73,8 +76,10 @@ public NetworkFailureHandler(
 
     /** Closes the specified channel after all queued write requests are flushed. */
     static void closeOnFlush(Channel channel) {
-        if (channel.isConnected()) {
+        if (channel.isConnected() && !channelsBeingClosed.containsKey(channel)) {
+            channelsBeingClosed.put(channel, channel);
             channel.write(ChannelBuffers.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE);
+            channelsBeingClosed.remove(channel);
```

Comment:
Shouldn't we remove it only when we actually close the channel? Something like:

```
    /** Closes the specified channel after all queued write requests are flushed. */
    static void closeOnFlush(Channel channel) {
        if (channel.isConnected() && !channelsBeingClosed.contains(channel)) {
            channelsBeingClosed.add(channel);
            channel.write(ChannelBuffers.EMPTY_BUFFER).addListener(CLOSE_WITH_BOOKKEEPING);
        }
    }

    private static final ChannelFutureListener CLOSE_WITH_BOOKKEEPING =
            future ->
                    future.getChannel()
                            .close()
                            .addListener(
                                    closedChannel ->
                                            channelsBeingClosed.remove(closedChannel.getChannel()));
```

Reply:
Thanks for the review @dawidwys!

Hmm.. I think the current approach is better because:

1) The code is less and also simpler to understand.

2) The approach makes less assumption in the sense that it just prevents closeOnFlush() from being called recursively (which we know will lead to live lock). 

In comparison, the alternative approach assumes that closeOnFlush() will never need to be called more than once for the same Channel), which may be true but is (IMO) not that obvious.

Is there any concern that we allow closeOnFlush() to be called multiple times for the same channel? If not, maybe it is better to take the simpler approach.

Reply:
nit: I see. Your explanation makes sense. Could we make it more explicit then that the only purpose of `channelsBeingClosed` is to prevent recursive call of `closeOnFlush`? E.g. could we move the variable right above the ` closeOnFlush`? 

Moreover how about we rephrase the comment to:

```
// The set of channels that are being closed in closeOnFlush. This is needed to prevent closing a channel recursively see FLINK-22085
```

Reply:
Actually @lindong28 I gave it a second thought and I think there is a problem with the current approach.

The call to `channel.write(ChannelBuffers.EMPTY_BUFFER)` is an async call. There is no guarantee it finishes when you call `channelsBeingClosed.remove(channel);`. Therefore you still may end up in a "recursive" call that you call `closeOnFlush` because of a previous `closeOnFlush` call.

We know that the call to `write` has finished only in the `ChannelFutureListener`  added via `addListener`.

Reply:
Thank you @dawidwys for catching this. Yep you are right. I have updated the PR according to the suggestions.

For now I still keep the `channelsBeingClosed` above the `NetworkFailureHandler` constructor instead of moving it right above the `closeOnFlush()`.  This is to follow the current Flink code style where all `static final` variables are above the class constructor.

Could you take another look?